saleor/account/throttling.py

```diff
@@ -0,0 +1,123 @@
+from datetime import timedelta
+from math import ceil
+from typing import Optional
+
+from django.core.cache import cache
+from django.core.exceptions import ValidationError
+from django.utils import timezone
+
+from ..core.utils import get_client_ip
+from . import models
+from .error_codes import AccountErrorCode
+from .utils import retrieve_user_by_email
+
+MIN_DELAY = 1
+MAX_DELAY = 3600
+ATTEMPT_COUNTER_EXPIRE_TIME = 7200
+
+
+def authenticate_with_throttling(request, email, password) -> Optional[models.User]:
+    ip = get_client_ip(request)
+    if not ip:
+        raise ValidationError(
+            {
+                "email": ValidationError(
+                    "Can't indentify requester IP address.",
+                    code=AccountErrorCode.UNKNOWN_IP_ADDRESS.value,
+                )
+            }
+        )
+
+    ip_key = get_cache_key_failed_ip(ip)
+    block_key = get_cache_key_blocked_ip(ip)
+
+    # block the IP address before the next attempt to prevent concurrent requests
+    if add_block(block_key, MIN_DELAY) is False:
+        next_attempt_time = cache.get(block_key) or timezone.now() + timedelta(
+            seconds=MIN_DELAY
+        )
```

Comment:
Here is a short period, between ADD and GET, when we could display shorter login delay time. IMO it is irrelevant